File: packages/react/test/patch.test.ts
import { describe, expect, it } from 'vitest';
import { diffText, formatPatch, parsePatch, type DiffResult, type DiffRow } from 'diffine-react';

/** A row written the way a reader would describe it, for readable assertions. */
function shape(row: DiffRow): string {
  const before = row.before ? `${row.before.index + 1}:${row.before.text}` : '·';
  const after = row.after ? `${row.after.index + 1}:${row.after.text}` : '·';

  return `${row.kind} ${before} | ${after}`;
}

const shapes = (result: DiffResult) => result.rows.map(shape);

/** A document of numbered lines, for a comparison with room between its changes. */
const numbered = (count: number, from = 1) =>
  Array.from({ length: count }, (_unused, index) => `line ${index + from}`).join('\n');

describe('formatPatch', () => {
  it('writes nothing at all for two documents that are the same', () => {
    expect(formatPatch(diffText('one\ntwo', 'one\ntwo'))).toBe('');
  });

  it('writes the headers, the hunk and the lines around it', () => {
    const patch = formatPatch(diffText('a\nb\nc', 'a\nB\nc'));

    expect(patch).toBe(
      ['--- before', '+++ after', '@@ -1,3 +1,3 @@', ' a', '-b', '+B', ' c', ''].join('\n')
    );
  });

  it('names each side where a name was given', () => {
    const patch = formatPatch(diffText('a', 'b'), {
      before: 'a/src/index.ts',
      after: 'b/src/index.ts'
    });

    expect(patch.split('\n').slice(0, 2)).toEqual(['--- a/src/index.ts', '+++ b/src/index.ts']);
  });

  it('leaves the count off a hunk that covers one line', () => {
    expect(formatPatch(diffText('a', 'b'), { context: 0 })).toContain('@@ -1 +1 @@');
  });

  it('writes the lines that come before an insertion as its position', () => {
    const patch = formatPatch(diffText('a\nb', 'a\nnew\nb'), { context: 0 });

    expect(patch).toContain('@@ -1,0 +2 @@');
  });

  it('keeps as many unchanged lines either side as `context` asks for', () => {
    const before = numbered(20);
    const after = before.replace('line 10', 'line ten');

    expect(formatPatch(diffText(before, after), { context: 1 })).toContain('@@ -9,3 +9,3 @@');
    expect(formatPatch(diffText(before, after), { context: 3 })).toContain('@@ -7,7 +7,7 @@');
  });

  it('writes two changes far apart as two hunks', () => {
    const before = numbered(40);
    const after = before.replace('line 5', 'line five').replace('line 30', 'line thirty');
    const patch = formatPatch(diffText(before, after));

    expect(patch.match(/^@@/gmu)).toHaveLength(2);
  });

  it('joins two changes whose context would touch into one hunk', () => {
    const before = numbered(20);
    const after = before.replace('line 8', 'line eight').replace('line 12', 'line twelve');
    const patch = formatPatch(diffText(before, after));

    expect(patch.match(/^@@/gmu)).toHaveLength(1);
  });

  it('writes everything that went out above everything that came in', () => {
    const patch = formatPatch(diffText('a\nb\nc', 'A\nB\nC'), { context: 0 });

    expect(patch.split('\n').slice(3, 9)).toEqual(['-a', '-b', '-c', '+A', '+B', '+C']);
  });
});

describe('parsePatch', () => {
  it('reads a file for each pair of header lines', () => {
    const patch = [
      'diff --git a/one.txt b/one.txt',
      'index 0000000..1111111 100644',
      '--- a/one.txt',
      '+++ b/one.txt',
      '@@ -1 +1 @@',
      '-a',
      '+b',
      'diff --git a/two.txt b/two.txt',
      '--- a/two.txt',
      '+++ b/two.txt',
      '@@ -1 +1 @@',
      '-c',
      '+d',
      ''
    ].join('\n');

    const files = parsePatch(patch);

    expect(files.map((file) => [file.before, file.after])).toEqual([
      ['a/one.txt', 'b/one.txt'],
      ['a/two.txt', 'b/two.txt']
    ]);
    expect(shapes(files[1].result)).toEqual(['replace 1:c | 1:d']);
  });

  it('numbers the lines from the hunk header rather than from the hunk', () => {
    const patch = ['--- a', '+++ b', '@@ -40,2 +50,2 @@', ' keep', '-old', '+new', ''].join('\n');

    expect(shapes(parsePatch(patch)[0].result)).toEqual([
      'equal 40:keep | 50:keep',
      'replace 41:old | 51:new'
    ]);
  });

  it('leaves a jump in the numbers where one hunk ends and the next begins', () => {
    const patch = [
      '--- a',
      '+++ b',
      '@@ -1 +1 @@',
      '-a',
      '+A',
      '@@ -40 +40 @@',
      '-b',
      '+B',
      ''
    ].join('\n');
    const { rows } = parsePatch(patch)[0].result;

    expect(rows[0].before?.index).toBe(0);
    expect(rows[1].before?.index).toBe(39);
  });

  it('marks the words inside a pair of changed lines, as the engine would', () => {
    const patch = ['--- a', '+++ b', '@@ -1 +1 @@', '-the quick fox', '+the slow fox', ''].join(
      '\n'
    );
    const [row] = parsePatch(patch)[0].result.rows;

    expect(row.before?.segments.map((piece) => piece.text)).toEqual(['the ', 'quick', ' fox']);
    expect(row.after?.segments.map((piece) => piece.kind)).toEqual(['equal', 'insert', 'equal']);
  });

  it('takes the options the engine takes', () => {
    const patch = ['--- a', '+++ b', '@@ -1 +1 @@', '-a b', '+a c', ''].join('\n');
    const [row] = parsePatch(patch, { inline: 'none' })[0].result.rows;

    expect(row.before?.segments).toEqual([]);
  });

  it('counts what the patch carried, and calls the result complete', () => {
    const patch = ['--- a', '+++ b', '@@ -1,3 +1,3 @@', ' keep', '-old', '+new', ' tail', ''].join(
      '\n'
    );
    const { stats, complete, before, after } = parsePatch(patch)[0].result;

    expect(stats).toEqual({ unchanged: 2, changed: 1, inserted: 0, deleted: 0 });
    expect(complete).toBe(true);
    expect(before).toEqual(['keep', 'old', 'tail']);
    expect(after).toEqual(['keep', 'new', 'tail']);
  });

  it('reads an empty line inside a hunk as an unchanged line', () => {
    const patch = ['--- a', '+++ b', '@@ -1,3 +1,3 @@', ' a', '', '-b', '+B', ''].join('\n');

    expect(shapes(parsePatch(patch)[0].result)).toEqual([
      'equal 1:a | 1:a',
      'equal 2: | 2:',
      'replace 3:b | 3:B'
    ]);
  });

  it('skips the marker for a file that does not end in a newline', () => {
    const patch = [
      '--- a',
      '+++ b',
      '@@ -1 +1 @@',
      '-a',
      '\\ No newline at end of file',
      '+b',
      '\\ No newline at end of file',
      ''
    ].join('\n');

    expect(shapes(parsePatch(patch)[0].result)).toEqual(['replace 1:a | 1:b']);
  });

  it('reads a patch that is nothing but hunks', () => {
    const files = parsePatch(['@@ -1 +1 @@', '-a', '+b', ''].join('\n'));

    expect(files).toHaveLength(1);
    expect(files[0].before).toBe('');
    expect(shapes(files[0].result)).toEqual(['replace 1:a | 1:b']);
  });

  it('finds nothing in text that is not a patch', () => {
    expect(parsePatch('just some prose\nover two lines')).toEqual([]);
  });

  it('reads a line that begins with three dashes as a line', () => {
    const patch = ['--- a', '+++ b', '@@ -1 +1 @@', '--- not a header', '+kept', ''].join('\n');

    expect(shapes(parsePatch(patch)[0].result)).toEqual(['replace 1:-- not a header | 1:kept']);
  });
});

describe('formatPatch and parsePatch together', () => {
  it('comes back to the same rows it started from', () => {
    const before = `${numbered(30)}\ntail`;
    const after = before.replace('line 4', 'line four').replace('line 25', '');
    const original = diffText(before, after);
    const patch = formatPatch(original);
    const [file] = parsePatch(patch);

    // Every changed row survives, and so does the context around it. What the
    // patch left out are unchanged rows the two documents agreed about.
    expect(shapes(file.result).filter((row) => !row.startsWith('equal'))).toEqual(
      shapes(original).filter((row) => !row.startsWith('equal'))
    );
    expect(file.result.changes).toHaveLength(original.changes.length);
  });

  it('writes the same patch again from what it read', () => {
    const before = numbered(30);
    const after = before.replace('line 4', 'line four').replace('line 25', 'line twenty-five');
    const patch = formatPatch(diffText(before, after));

    expect(formatPatch(parsePatch(patch)[0].result)).toBe(patch);
  });
});

File: packages/flutter/test/diff/patch_test.dart
import 'package:diffine/diffine.dart';
import 'package:flutter_test/flutter_test.dart';

const String _before = 'one\ntwo\nthree\nfour\nfive\nsix\nseven\n';
const String _after = 'one\ntwo\nthree\nFOUR\nfive\nsix\nseven\n';

List<String> lines(String patch) => patch.split('\n');

void main() {
  group('formatPatch', () {
    test('writes nothing at all for two documents that are the same', () {
      expect(formatPatch(diffText(_before, _before)), '');
    });

    test('writes the headers, the hunk and the lines around it', () {
      expect(lines(formatPatch(diffText(_before, _after))), <String>[
        '--- before',
        '+++ after',
        '@@ -1,7 +1,7 @@',
        ' one',
        ' two',
        ' three',
        '-four',
        '+FOUR',
        ' five',
        ' six',
        ' seven',
        '',
      ]);
    });

    test('names each side where a name was given', () {
      final String patch = formatPatch(
        diffText(_before, _after),
        const DiffPatchOptions(before: 'a/lib/main.dart', after: 'b/lib/main.dart'),
      );

      expect(lines(patch).sublist(0, 2), <String>['--- a/lib/main.dart', '+++ b/lib/main.dart']);
    });

    test('leaves the count off a hunk that covers one line', () {
      expect(formatPatch(diffText('a', 'b')), contains('@@ -1 +1 @@'));
    });

    test('writes the lines that come before an insertion as its position', () {
      expect(formatPatch(diffText('', 'a\n')), contains('@@ -0,0 +1 @@'));
    });

    test('keeps as many unchanged lines either side as `context` asks for', () {
      final String patch = formatPatch(
        diffText(_before, _after),
        const DiffPatchOptions(context: 1),
      );

      expect(lines(patch).sublist(2, 6), <String>['@@ -3,3 +3,3 @@', ' three', '-four', '+FOUR']);
    });

    test('writes two changes far apart as two hunks', () {
      final String before = List<String>.generate(40, (int i) => 'line $i').join('\n');
      final String after = before
          .replaceFirst('line 2', 'LINE 2')
          .replaceFirst('line 30', 'LINE 30');
      final String patch = formatPatch(diffText(before, after));

      expect(RegExp('@@').allMatches(patch), hasLength(4));
    });

    test('joins two changes whose context would touch into one hunk', () {
      final String before = List<String>.generate(12, (int i) => 'line $i').join('\n');
      final String after = before.replaceFirst('line 4', 'LINE 4').replaceFirst('line 7', 'LINE 7');
      final String patch = formatPatch(diffText(before, after));

      expect(RegExp('@@').allMatches(patch), hasLength(2));
    });

    test('writes everything that went out above everything that came in', () {
      final String patch = formatPatch(diffText('a\nb\nc\n', 'A\nB\nC\n'));

      expect(lines(patch).sublist(3, 9), <String>['-a', '-b', '-c', '+A', '+B', '+C']);
    });
  });

  group('parsePatch', () {
    test('reads a file for each pair of header lines', () {
      const String patch = '''
--- a/one
+++ b/one
@@ -1 +1 @@
-a
+A
--- a/two
+++ b/two
@@ -1 +1 @@
-b
+B
''';
      final List<DiffPatchFile> files = parsePatch(patch);

      expect(files, hasLength(2));
      expect(files.first.before, 'a/one');
      expect(files[1].after, 'b/two');
    });

    test('numbers the lines from the hunk header rather than from the hunk', () {
      const String patch = '''
--- a
+++ b
@@ -10,2 +10,2 @@
-old
+new
''';
      final DiffResult result = parsePatch(patch).single.result;

      expect(result.rows.first.before!.index, 9);
      expect(result.rows.first.after!.index, 9);
    });

    test('leaves a jump in the numbers where one hunk ends and the next begins', () {
      const String patch = '''
--- a
+++ b
@@ -1,1 +1,1 @@
-a
+A
@@ -40,1 +40,1 @@
-b
+B
''';
      final DiffResult result = parsePatch(patch).single.result;

      expect(result.rows.first.before!.index, 0);
      expect(result.rows[1].before!.index, 39);
    });

    test('marks the words inside a pair of changed lines, as the engine would', () {
      const String patch = '''
--- a
+++ b
@@ -1 +1 @@
-the quick brown fox
+the slow brown fox
''';
      final DiffRow row = parsePatch(patch).single.result.rows.first;

      expect(row.before!.segments, isNotEmpty);
      expect(
        row.before!.segments.map((DiffSegment piece) => piece.text).join(),
        'the quick brown fox',
      );
    });

    test('takes the options the engine takes', () {
      const String patch = '''
--- a
+++ b
@@ -1 +1 @@
-the quick fox
+the slow fox
''';
      final DiffRow row = parsePatch(
        patch,
        const DiffOptions(inline: DiffInlineMode.none),
      ).single.result.rows.first;

      expect(row.before!.segments, isEmpty);
    });

    test('counts what the patch carried, and calls the result complete', () {
      const String patch = '''
--- a
+++ b
@@ -1,3 +1,3 @@
 keep
-old
+new
''';
      final DiffResult result = parsePatch(patch).single.result;

      expect(result.stats.unchanged, 1);
      expect(result.stats.changed, 1);
      expect(result.complete, isTrue);
    });

    test('reads an empty line inside a hunk as an unchanged line', () {
      const String patch = '--- a\n+++ b\n@@ -1,3 +1,3 @@\n keep\n\n-old\n+new\n';
      final DiffResult result = parsePatch(patch).single.result;

      expect(result.before, <String>['keep', '', 'old']);
    });

    test('skips the marker for a file that does not end in a newline', () {
      const String patch = '''
--- a
+++ b
@@ -1 +1 @@
-old
\\ No newline at end of file
+new
''';
      final DiffResult result = parsePatch(patch).single.result;

      expect(result.before, <String>['old']);
      expect(result.after, <String>['new']);
    });

    test('reads a patch that is nothing but hunks', () {
      final DiffResult result = parsePatch('@@ -1 +1 @@\n-a\n+b\n').single.result;

      expect(result.rows, hasLength(1));
    });

    test('finds nothing in text that is not a patch', () {
      expect(parsePatch('nothing here\nor here\n'), isEmpty);
    });

    test('comes back to the same rows it started from', () {
      final DiffResult worked = diffText(_before, _after);
      final DiffResult read = parsePatch(formatPatch(worked)).single.result;

      expect(
        read.rows.map((DiffRow row) => '${row.kind.name} ${row.before?.text}').toList(),
        worked.rows.map((DiffRow row) => '${row.kind.name} ${row.before?.text}').toList(),
      );
    });

    test('writes the same patch again from what it read', () {
      final String patch = formatPatch(diffText(_before, _after));
      final DiffPatchFile file = parsePatch(patch).single;

      expect(
        formatPatch(file.result, DiffPatchOptions(before: file.before, after: file.after)),
        patch,
      );
    });
  });
}
